### PR TITLE
Add warning for duplicate pip requirements

### PR DESCRIPTION
**Pull request by BenWilson2**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Adds a warning if duplicate pip requirements are applied by a user specifying conflicting base package names in extra_pip_requirements

## How is this patch tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by BenWilson2** [comment]
@dbczumar - is this what we're thinking of for warning users about dupes?
cc @ankit-db

File: mlflow/utils/environment.py [hunk 1]
```diff
@@ -500,6 +501,19 @@ def _process_pip_requirements(
     return conda_env, pip_reqs, constraints
 
 
+def _check_for_duplicate_requirements(requirements):
```
Comment (harupy) [inline mlflow/utils/environment.py:504]:
```suggestion
def _find_duplicate_requirements(requirements):
```

maybe `find` is more straightforward?

File: mlflow/utils/environment.py [hunk 2]
```diff
@@ -500,6 +501,19 @@ def _process_pip_requirements(
...
+    """
+    Checks if duplicate base package requirements are specified in any list of requirements
+    and returns the list of duplicate base package names
+    """
+    base_package_names = [
+        Requirement(package).name for package in requirements if not package.startswith("-")
```
Comment (harupy) [inline mlflow/utils/environment.py:510]: Does `Requirement(package)` work for requirements like `git+https://github.com/...`? I remember it doesn't.

Comment (harupy) [reply mlflow/utils/environment.py:510]: > Does `Requirement(package)` work for requirements like `git+https://github.com/...`? I remember it doesn't. If it doesn't, we need to suppress an exception.

Comment (BenWilson2) [reply mlflow/utils/environment.py]: Updated with handling and updated the test with all supported methods of installing packages (or restricting installation) in python as of the most recent version of pip.

File: mlflow/utils/environment.py (hunk 1 above)
Comment (BenWilson2) [reply mlflow/utils/environment.py]: good point :)

File: mlflow/utils/environment.py [hunk 3]
```diff
@@ -500,6 +501,29 @@ def _process_pip_requirements(
...
+    """
+    Checks if duplicate base package requirements are specified in any list of requirements
+    and returns the list of duplicate base package names.
+    Note that git urls and paths to local files are not being considered for duplication checking.
+    """
+    base_package_names = []
+
+    for package in requirements:
+        if package.startswith("-") or package.startswith("/") or "://" in package:
+            # Skip flags, local file paths, and urls
```
Comment (dbczumar) [inline mlflow/utils/environment.py:514]: Do we need this? Seems like try / catch requirement parsing should work :)

File: mlflow/transformers.py [hunk 4]
```diff
@@ -505,6 +506,14 @@ def save_model(
     else:
         conda_env, pip_requirements, pip_constraints = _process_conda_env(conda_env)
 
+    if duplicates := _find_duplicate_requirements(pip_requirements):
+        _logger.warning(
+            "Duplicate packages are present within the pip requirements. Duplicate packages: "
+            f"{duplicates}. Please manually specify the requirements by using the "
+            "`pip_requirements` argument in order to prevent unexpected installation "
+            "issues for this model."
+        )
```
Comment (dbczumar) [inline mlflow/transformers.py:515]: I think we should apply this to all flavors, not just transformers. Can we file a follow-up JIRA for next sprint?

File: mlflow/utils/environment.py (hunk 3 above)
Comment (BenWilson2) [reply mlflow/utils/environment.py:514]: wow. yeah. 
Also, the private repo declarations actually get captured by the Requirement library. That's one additional dupe check! Thanks for this comment :D

File: tests/utils/test_environment.py [hunk 5]
```diff
@@ -326,3 +327,31 @@ def make_conda_env(pip_deps):
...
+        "pytest-cov; python_version < '3.8'",  # conditional requirements
+        "fastapi @ file:///local/path/to/numpy-0.22.0-py3-none-any.whl",  # wheel from a local file
+        "-c constraints.txt",  # constraint file
+        "--no-binary :all:",  # no binary packages, source code only
+        "--prefer-binary",  # prefer binary packages over source code
+        "numpy<1.24.0",  # maximum version
+        "numpy",  # any version
+    ]
+    evaluation = _find_duplicate_requirements(packages)
+    assert sorted(evaluation) == ["numpy", "package"]
```
Comment (dbczumar) [inline tests/utils/test_environment.py]: Can we test that a warning is emitted when we call `mlflow.transformers.log_model()`? That way, we know `_find_duplicate_requirements` is called properly by `mlflow.transformers.log_model()` :)

Comment (dbczumar) [reply tests/utils/test_environment.py]: (Feel free to mock `logger.warning` :D )

**Pull request closed without merging**